Draft: Use abi3

ChristosT opened the pull request:
This is an attempt to use stable python API only. Using the stable API  will allow to create a single  *.abi3.so library for all versions of python 3.x .

Few conversions to stable API (non-checking vs checking version of common calls) already took place. The current build status of the python modules using `0x030B0000` is:

- [x]  `conduit_python`
  - [x] `PyObject_Print` - use just `printf` since we create the string ourselves
  - [x] `Py_buffer API` (stable only for 3.11+)
  - [x] `PyUnicode_FromKindAndData(PyUnicode_4BYTE_KIND,` - use `PyUnicode_Decode`
  - [x] static typed defined through `tp_dict` need to become heap allocated through `PyType_NewFromSpec`
- [x]  `conduit_relay_python` 
- [x] `conduit_relay_python_io_blueprint`
- [x] `conduit_relay_io_python`
	- [x] static typed defined through `tp_dict` need to become heap allocated through `PyType_NewFromSpec`
- [x] `conduit_relay_mpi_python`
	- [x] static typed defined through `tp_dict` need to become heap allocated through `PyType_NewFromSpec`
- [x] `conduit_relay_web_python`
	- [x] static typed defined through `tp_dict` need to become heap allocated through `PyType_NewFromSpec`

- [x] `conduit_blueprint_python`
- [x] `conduit_blueprint_mcarray_python`
- [x] `conduit_blueprint_mcarray_examples_python`
- [x] `conduit_blueprint_mesh_python`
- [x] `conduit_blueprint_mesh_examples_python`
- [x] `conduit_blueprint_table_python`
- [x] `conduit_blueprint_table_examples_python`

ChristosT committed:
Use Limited API: PyBytesAS_STRING -> PyBytes_AsString

ChristosT committed:
Use Limited API: PyList_Size->PyList_GET_SIZE

ChristosT committed:
Use Limited API: PyTuple_GET_ITEM -> PyTuple_GetItem

ChristosT committed:
Use Limited API: PyTuple_GET_SIZE -> PyTuple_Size

ChristosT committed:
Use Limited API: PyList_GET_ITEM -> PyList_GetItem

ChristosT committed:
Use Limited API: PyFloat_AS_DOUBLE -> PyFloat_AsDouble

ChristosT committed:
BUGFIX: PyUnicode_Decode expects size in bytes

while PyUnicode_FromKindAndData expects size of units of bytes

ChristosT committed:
Provide alternative implementation for PyUnicode_FromKindAndData

ChristosT committed:
Simplify PyConduit_Node_print_detailed

Just use std::fprintf, no need to construct a python string just to print it to stdout.
Also, PyObject_Print is not part of the Stable API.

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -100,7 +100,11 @@ PyString_AsString(PyObject *py_obj)
                                                           "strict"); // Owned reference
         if(temp_bytes != NULL)
         {
+            #ifdef Py_LIMITED_API
+            res = _conduit_strdup(PyBytes_AsString(temp_bytes));
+            #else
             res = _conduit_strdup(PyBytes_AS_STRING(temp_bytes));
+            #endif
```
It's probably better to just always use the ABI-stable code here. This also means that the non-ABI-stable builds test the same codepaths.

ChristosT committed:
Use PyType_GenericAlloc instead of tp_alloc

In the limited python API the PyTypeObject is considered an opaque pointer
we use PyType_GenericAlloc instead which is part of the Stable ABI.

ChristosT committed:
Use PyType_GetSlot to access tp_free

needed since PyTypeObject is an opaque struct in the limited api.

mathstuf commented on src/libs/CMakeLists.txt:
```diff
@@ -2,6 +2,17 @@
+#set(CONDUIT_PYTHON_LIMITED_API)
+set(CONDUIT_PYTHON_LIMITED_API 0x030A0000)
+
+if(CONDUIT_PYTHON_LIMITED_API)
```
This variable should be guarded by a version check of Python itself that it is of at least 3.10 as well (or whatever the limited API version ends up being).

ChristosT replied:
good point !

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -183,6 +198,35 @@ PyUnicode_From_UTF32_Unicode_Buffer(const char *unicode_buffer,
+};
+
+//---------------------------------------------------------------------------//
+#if defined(IS_PY3K)
```
Is Python2 really still supported?

ChristosT replied:
This is original code that was brought from bellow around line 8128

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -271,7 +315,16 @@ PyConduit_DataType_new(PyTypeObject* type,
         return (NULL);
     }
 
+    #ifdef Py_LIMITED_API
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    PyConduit_DataType *self = (PyConduit_DataType*)tp_alloc(type, 0);
+    #else
     PyConduit_DataType *self = (PyConduit_DataType*)type->tp_alloc(type, 0);
+    #endif
```
You can probably just use [`PyType_GenericAlloc`](https://docs.python.org/3/c-api/type.html#c.PyType_GenericAlloc) instead.

ChristosT replied:
Thanks I wasn't aware of this function .

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -467,7 +520,12 @@ PyConduit_DataType_init(PyConduit_DataType* self,
 PyConduit_DataType_dealloc(PyConduit_DataType *self)
 {
+    #ifdef Py_LIMITED_API
+    freefunc tp_free = ((freefunc)PyType_GetSlot(Py_TYPE(self), Py_tp_free));
```
Alas, this does not seem to have an ABI-stable wrapper :/ .

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -3210,20 +3284,38 @@ static PyTypeObject PyConduit_DataType_TYPE = {
-PyConduit_DataType_Python_Create()
+PyConduit_DataType_Python_Create(struct module_state* state)
 {
+    #ifdef Py_LIMITED_API
+    assert(state != NULL);
+    PyTypeObject* type = state->PyConduit_DataType_TYPE;
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    return (PyConduit_DataType*)tp_alloc(type, 0);
+    #else
+    (void*)(state);
     PyTypeObject* type = (PyTypeObject*)&PyConduit_DataType_TYPE;
     return (PyConduit_DataType*)type->tp_alloc(type,0);
+    #endif
```
`PyType_GenericAlloc`?

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -3241,7 +3333,16 @@ PyConduit_Generator_new(PyTypeObject *type,
                         PyObject*, // args -- unused
                         PyObject*) // kwds -- unused
 {
+    #ifdef Py_LIMITED_API
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    PyConduit_Generator *self = (PyConduit_Generator*)tp_alloc(type, 0);
+    #else
     PyConduit_Generator *self = (PyConduit_Generator*)type->tp_alloc(type, 0);
+    #endif
```
`PyType_GenericAlloc`?

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -4551,7 +4718,16 @@ PyConduit_NodeIterator_new(PyTypeObject *type,
         return (NULL);
     }
     
+    #ifdef Py_LIMITED_API
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    PyConduit_DataType *self = (PyConduit_DataType*)tp_alloc(type, 0);
+    #else
     PyConduit_DataType *self = (PyConduit_DataType*)type->tp_alloc(type, 0);
+    #endif
```
`PyType_GenericAlloc`?

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -4901,7 +5082,16 @@ static PyConduit_NodeIterator *
 PyConduit_NodeIterator_Python_Create()
 {
     PyTypeObject* type = (PyTypeObject*)&PyConduit_NodeIterator_TYPE;
+    #ifdef Py_LIMITED_API
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    return (PyConduit_NodeIterator*)tp_alloc(type, 0);
+    #else
     return (PyConduit_NodeIterator*)type->tp_alloc(type,0);
+    #endif
```
`PyType_GenericAlloc`?

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -5025,7 +5215,17 @@ PyConduit_Node_new(PyTypeObject* type,
     }
 
+
+    #ifdef Py_LIMITED_API
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    PyConduit_Node* self = (PyConduit_Node*)tp_alloc(type, 0);
+    #else
     PyConduit_Node* self = (PyConduit_Node*)type->tp_alloc(type, 0);
+    #endif
```
`PyType_GenericAlloc`?

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -6041,11 +6246,15 @@ PyConduit_Node_print_detailed(PyConduit_Node *self)
     std::ostringstream oss;
     self->node->to_string_stream(oss,"conduit_json");
+    #ifdef Py_LIMITED_API
+    std::fprintf(stdout,"%s",oss.str().c_str());
```
It seems that this can just always be used…the Python juggling seems excessive anyways.

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -7235,7 +7456,16 @@ PyConduit_Schema_Python_Wrap(Schema *schema, int python_owns)
 {
     PyTypeObject *type = (PyTypeObject*)&PyConduit_Schema_TYPE;
 
+    #ifdef Py_LIMITED_API
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    PyConduit_Schema* retval = (PyConduit_Schema*)tp_alloc(type, 0);
+    #else
     PyConduit_Schema *retval = (PyConduit_Schema*)type->tp_alloc(type, 0);
+    #endif
```
`PyType_GenericAlloc`?

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -7261,7 +7491,17 @@ static PyObject *
 {
     PyTypeObject* type = (PyTypeObject*)&PyConduit_Node_TYPE;
+
+    #ifdef Py_LIMITED_API
+    allocfunc tp_alloc = (allocfunc)PyType_GetSlot(type, Py_tp_alloc);
+    if(!tp_alloc)
+    {
+      return NULL;
+    }
+    PyConduit_Node* retval = (PyConduit_Node*)tp_alloc(type, 0);
+    #else
     PyConduit_Node* retval = (PyConduit_Node*)type->tp_alloc(type, 0);
+    #endif
```
`PyType_GenericAlloc`?

mathstuf commented on src/libs/conduit/python/conduit_python.cpp:
```diff
@@ -7286,7 +7526,11 @@ PyConduit_Node_Set_From_Python_List(Node &node,
     // a numeric case, or a more general case
     
+    #ifdef Py_LIMITED_API
+    Py_ssize_t list_size = PyList_Size(value);
```
Same here; just unconditionally use the stable ABI (and elsewhere).

mathstuf reviewed (changes requested):
I think a number of the commits here could be split out into their own PR to help focus the ABI3 parts.

ChristosT commented:
> I think a number of the commits here could be split out into their own PR to help focus the ABI3 parts.

Yes, the rebase mixed up the history. I will remove those and maybe I can keep 6c3db9f52c5e2e6144ee4c3a12b05d2f10d94705 separate.

ChristosT committed:
cmake: create interface library that holds Py_LIMITED symbol

The new cmake option CONDUIT_PYTHON_USE_LIMITED_API can be used to toggle the use of limited API.
The Python version is set to 3.11 since it is required for the Py_Buffer objects.

ChristosT committed:
conduit_python: Include missing header for Python 3.11

ChristosT committed:
conduit_python: Prepare module for heap allocated types

*_TYPE objects will be saved in module state and initialized in
PyInit_conduit_python method.

ChristosT committed:
Convert PyConduit_DataType_TYPE to heap allocated type

ChristosT committed:
Convert PyConduit_Generator_TYPE to heap allocated type

ChristosT committed:
Convert PyConduit_Schema_TYPE to heap allocated type

ChristosT committed:
Convert PyConduit_NodeIterator_TYPE to heap allocated type

ChristosT committed:
Convert PyConduit_Node_TYPE to heap allocated type

ChristosT committed:
Convert PyConduit_Endianess_TYPE to heap allocated type

ChristosT commented:
At this stage the `conduit_python` library compiles and all python tests run successfully when  setting the limited api to 3.11 .

cyrush committed:
0.8.7 release prep (#1096)

* fix test

* refactor NDIndex into blueprint::ndarray namespace

* cleanup of old namespace alias to avoid confusing with new object

* small changes

* warnings pass for a few int types

* warnings pass avoid pragma msgs

* resolve pessimizing-move (allow copy elison)

* warnings pass remove extra include

* update changelog

* docs updates

* warnings pass

* remove stale host configs and docs ref

ChristosT committed:
conduit_relay_io_python: Prepare module for heap allocated types

ChristosT committed:
Convert PyRelay_IOHandle_TYPE to heap allocated type

ChristosT committed:
conduit_replay_web_python: prepare for heap allocated types

ChristosT committed:
Convert PyRelay_WebServer_TYPE to heap allocated type

ChristosT committed:
Convert PyRelay_WebSocket_TYPE to heap allocated type

ChristosT committed:
conduit_relay_mpi_python: Prepare for heap allocated types

ChristosT committed:
Convert PyRelay_MPI_Request_TYPE to heap-allocated type

ChristosT commented:
@mathstuf  all types can  now be heap allocated and the library can be built targeting the ABI of 3.11. 
python tests pass locally on and `abi3audit` reports zero violations when creating a wheel. 

Let me know if you have any further comments before opening an upstream pull request.

ChristosT commented:
closing since changes have been merged upstream
